node/node.go

```diff
@@ -363,66 +367,313 @@ func (node *Node) AddPendingReceipts(receipts *types.CXReceiptsProof) {
 		Msg("Got ONE more receipt message")
 }
 
+type withError struct {
+	err     error
+	payload interface{}
+}
+
+var (
+	errNotRightKeySize = errors.New("key received over wire is wrong size")
+	errNoSenderPubKey  = errors.New("no sender public BLS key in message")
+	errWrongShardID    = errors.New("wrong shard id")
+)
+
+// validateShardBoundMessage validate consensus message
+// validate shardID
+// validate public key size
+// verify message signature
+func (node *Node) validateShardBoundMessage(
+	ctx context.Context, payload []byte,
+) (*msg_pb.Message, *bls.PublicKey, error) {
+	var (
+		m         msg_pb.Message
+		senderKey *bls.PublicKey
+	)
+	entryTime := time.Now()
+	defer utils.Logger().Debug().Str("cost", time.Now().Sub(entryTime).String()).Msg("[cost:validate_shard_bound_message]")
+
+	if err := protobuf.Unmarshal(payload, &m); err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+	var senderPubKeyViaWire []byte
+	maybeCon, maybeVC := m.GetConsensus(), m.GetViewchange()
+
+	if maybeCon != nil {
+		if maybeCon.ShardId != node.Consensus.ShardID {
+			return nil, nil, errors.WithStack(errWrongShardID)
+		}
+		senderPubKeyViaWire = maybeCon.GetSenderPubkey()
+	} else if maybeVC != nil {
+		if maybeVC.ShardId != node.Consensus.ShardID {
+			return nil, nil, errors.WithStack(errWrongShardID)
+		}
+		senderPubKeyViaWire = maybeVC.GetSenderPubkey()
+	} else {
+		return nil, nil, errors.WithStack(errNoSenderPubKey)
+	}
+
+	if len(senderPubKeyViaWire) != shard.PublicKeySizeInBytes {
+		return nil, nil, errors.WithStack(errNotRightKeySize)
+	}
+
+	key, err := bls_cosi.BytesToBLSPublicKey(
+		senderPubKeyViaWire,
+	)
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+	senderKey = key
+
+	if !node.Consensus.IsValidatorInCommittee(senderKey) {
```

Comment:
Is there a data race issue here?

Reply:
why?  this is read-only access

Reply:
The decider can be changed: consensus/consensus_service.go:393

Reply:
The Decider object can be changed: consensus/consensus_service.go:393